### PR TITLE
Add chained verification of certificate signatures.

### PR DESCRIPTION
**Pull request by thmsbinder**
Ref https://github.com/project-oak/oak/issues/4074

### CODE REVIEW

File: oak_attestation_verification/src/lib.rs [hunk 1]
```diff
@@ -15,6 +15,7 @@
 //
 
 #![cfg_attr(not(test), no_std)]
+#![feature(let_chains)]
```
Comment (ipetr0v) [inline oak_attestation_verification/src/lib.rs:18]: What is this feature?
Do we provide it? Or it comes from a dependency?

File: oak_attestation_verification/src/util.rs [hunk 2]
```diff
@@ -85,15 +83,15 @@ pub fn equal_keys(public_key_a: &[u8], public_key_b: &[u8]) -> anyhow::Result<bo
...
 
 /// Verifies the signature over the contents using the public key.
-pub fn verify_signature(
+pub fn verify_signature_raw(
     signature: &[u8],
     contents: &[u8],
     public_key: &[u8],
 ) -> anyhow::Result<()> {
-    let signature = Signature::from_der(signature).context("invalid ASN.1 signature")?;
+    let sig = Signature::from_der(signature).context("invalid ASN.1 signature")?;
```
Comment (ipetr0v) [inline oak_attestation_verification/src/util.rs:91]: nit: Why did it shorten?

File: oak_attestation_verification/src/verifier.rs [hunk 3]
```diff
@@ -73,8 +87,30 @@ pub fn verify_internal(
     }
 }
 
-fn verify_certificate_chain(_evidence: &Evidence) -> anyhow::Result<()> {
-    anyhow::bail!("Needs implementation")
+fn verify_certificate_chain(evidence: &Evidence) -> anyhow::Result<()> {
```
Comment (ipetr0v) [inline oak_attestation_verification/src/verifier.rs:90]: I think we could rename it to `verify_dice_certificate_chain` to be explicit

File: oak_attestation_verification/src/verifier.rs [hunk 4]
```diff
@@ -90,7 +126,7 @@ fn verify_oak_containers(
     _endorsements: &OakContainersEndorsements,
     _reference_values: &OakContainersReferenceValues,
 ) -> anyhow::Result<()> {
-    anyhow::bail!("Needs implementation")
+    Ok(()) // Work in progress
```
Comment (ipetr0v) [inline oak_attestation_verification/src/verifier.rs:129]: Could you instead add a `// TODO(#4074): ...`

File: oak_attestation_verification/src/verifier.rs [hunk 5]
```diff
@@ -100,3 +136,23 @@ fn verify_cb(
 ) -> anyhow::Result<()> {
     anyhow::bail!("Needs implementation")
 }
+
+/// Extracts the certified public key from the claims set of a certificate.
+/// NB: Copied from oak_dice crate.
```
Comment (ipetr0v) [inline oak_attestation_verification/src/verifier.rs:141]: nit: What is NB?

**Review by conradgrobler — COMMENTED** [review]
Why are so many things copied, rather than referencing the oak_dice crate from here? I think maintaining multiple copies of this logic is error-prone and unnecessary.

**Review by ipetr0v — APPROVED** [review]

File: oak_attestation_verification/src/lib.rs (hunk 1 above)
Comment (thmsbinder) [reply oak_attestation_verification/src/lib.rs]: It was needed to get code copied from oak_dice going. But we don't copy anymore, so this is gone.

File: oak_attestation_verification/src/verifier.rs (hunk 4 above)
Comment (thmsbinder) [reply oak_attestation_verification/src/verifier.rs:124]: Done.

File: oak_attestation_verification/src/util.rs (hunk 2 above)
Comment (thmsbinder) [reply oak_attestation_verification/src/util.rs:90]: In order not to shadow the function parameter. It is probably technically OK but seems careless.

File: oak_attestation_verification/src/verifier.rs (hunk 3 above)
Comment (thmsbinder) [reply oak_attestation_verification/src/verifier.rs:83]: Going with verify_dice_chain.

**Comment by ipetr0v** [comment]
> Removed code copied from oak_dice and added a dependency instead.

`oak_dice` doesn't have the logic for generating attestation reports, right? Only the DICE Rust structs?

**Comment by thmsbinder** [comment]
> > Removed code copied from oak_dice and added a dependency instead.
> 
> `oak_dice` doesn't have the logic for generating attestation reports, right? Only the DICE Rust structs?

Yes. oak_dice is data structures + utilities for e.g. oak_remote_attestation.  We could consider it as the common (non-public) part for attestation-related things.

**Review by conradgrobler — APPROVED** [review]
Thanks

File: oak_attestation_verification/src/verifier.rs [hunk 6]
```diff
@@ -73,8 +80,33 @@ pub fn verify_internal(
...
+        cose_key_to_verifying_key(&cose_key).map_err(|msg| anyhow::anyhow!(msg))?;
+
+    for layer in evidence.layers.iter() {
+        let cert = coset::CoseSign1::from_slice(&layer.eca_certificate)
+            .map_err(|_cose_err| anyhow::anyhow!("could not parse certificate"))?;
+        cert.verify_signature(ADDITIONAL_DATA, |signature, contents| {
+            let sig = Signature::from_slice(signature)?;
+            verifying_key.verify(contents, &sig)
+        })?;
+        let payload = cert.payload.ok_or(anyhow::anyhow!("no cert payload"))?;
```
[the remark this answers is not in the document]
Comment (thmsbinder) [reply oak_attestation_verification/src/verifier.rs:101]: Done.

**Pull request merged**